README.md
# Client HTTP/HTTPS request in node

[![NPM version](https://badge.fury.io/js/client-http.png)](http://badge.fury.io/js/client-http)

## Feature List

1. Automatic proxy setting, if you set the proxy in your environment variable(http_proxy), it will be auto detected and used in your any HTTP request, also it will be used in your HTTPS request by HTTPS Tunnel(denpency by node-tunnel) to make HTTPS request through HTTP. Optional, you can set proxy manually by call setProxy(proxy).

2. Make your HTTP/HTTPS request with cookie, cookie will be automatic processed int any HTTP/HTTPS request and you can get the cookie value in response.

3. HTTP redirect handling, HTTP redirect(301/302) will be automatic handled in you HTTP/HTTPS request.

## Example
    
    var http = require('client-http');

    http.get("http://www.google.com/", function(data){
        data && console.log(data);
    });

## Installation

    $ npm install client-http

## Usages

### HTTPS request

    var http = require('client-http');

    http.get("https://www.google.com/", function(data){
        data && console.log(data);
    });

### Read cookie from response

    var http = require('client-http');

    http.get("http://www.google.com/", function(data, err, cookie){
        !err && console.log(cookie);
    });

### Read header from response

    var http = require('client-http');

    http.get("http://www.google.com/", function(data, err, cookie, headers){
        !err && console.log(headers);
    });

### HTTP Post

    var http = require('client-http');

    http.request("http://www.snee.com/xml/crud/posttest.cgi", function(data){
        data && console.log(data);
    }, "fname=hello&lname=world");

### HTTP request with custom header

    var http = require('client-http');

    // we use iPhone as our user-agent and see what happen
    http.request("http://twitter.com/", function(data){
        data && console.log(data);
    }, null, {"User-Agent": "iPhone"});

### HTTP request with manually proxy set

    var http = require('client-http');
    
    http.setProxy('http://127.0.0.1:8080');
    http.get("https://www.google.com/", function(data){
        data && console.log(data);
    });

### HTTP request with manually timeout set

    var http = require('client-http');
   
    http.setTimeout(5000);
    http.get("https://www.google.com/", function(data){
        data && console.log(data);
    });


CHANGELOG.md
# Changelog

 - 0.0.1 (07/23/2012)
   - first release.

 - 0.0.2 (01/29/2013)
   - fix bug: callback is not called if http return buffer is empty

 - 0.0.3 (02/25/2013)
   - fix bug: can not request through non-standard http/https port, such as non 80 or non 443 port
   - enhancement: always return server response content, even if return status code is not 200

 - 0.0.4 (02/25/2013)
   - fix bug: emergency fix of version 0.0.3

 - 0.0.5 (06/25/2013)
   - enhancement: let you read response header in callback

 - 0.0.6 (02/04/2014)
   - enhancement: Allow set proxy manually by call setProxy 
 
 - 0.0.7 (05/30/2014)
   - enhancement: Allow set timeout(in milliseconds) before make http request, default timeout is 1 minutes


test/http-base.js
var http = require('../index.js');

http.get("http://www.google.com/", function(data, err){
    data && console.log(data);
    err && console.log(err);
});


test/http-post.js
var http = require('../index.js');

http.request("http://httpbin.org/post", function(data, err){
    data && console.log(data);
    err && console.log(err);
}, "fname=hello&lname=world", {"Content-Type":"application/x-www-form-urlencoded"});


test/https-base.js
var http = require('../index.js');

http.get("https://www.google.com/", function(data, err){
    data && console.log(data);
    err && console.log(err);
});


lib/clienthttp.js
var http = require('http'),
    https = require('https'),
    tunnel = require('tunnel'),
    url = require('url');
    
var _cookie = {};
var _proxy;
var _timeout = 60000;

exports.request   = httpRequest;
exports.get       = httpGet;
exports.setProxy  = function(proxy){
    _proxy = proxy;
};
exports.setTimeout = function(timeout){
    _timeout = timeout;
};

function httpRequest(rawUrl, callback, data, headers){ // data and custom header parameter is optional
    var options = genRequestOptions(rawUrl, function(err){
        err && callback && callback(null, err);
    }, data, headers);
    if (!options){
        return;
    }
    var req;
    if (options._protocol=="https" && !options.agent){
        req = https.request(options, function(res){
            handleResponese(res, callback, data, headers);
        });
    }else{
        req = http.request(options, function(res){
            handleResponese(res, callback, data, headers);
        });
    }
    req.on('socket', function(socket){
      socket.setTimeout(_timeout);
      socket.on('timeout', function(){
        req.abort();
      });
    });
    data && req.write(data);
    req.on("error", function(err){
        callback && callback(null,err);
    });
    req.end();
}

function httpGet(rawUrl, callback){
    httpRequest(rawUrl, callback);
}

function handleResponese(res, callback, data, headers){
    res.setEncoding("utf8");
    setCookieByHttpRes(res);
    var buffer = "";
    res.on("data", function(chunk){
        if (chunk && chunk.length>0){
            buffer += chunk;
        }
    });
    res.on("end", function(){
        if (res.statusCode==200){
            callback && callback(buffer, null, _cookie, res.headers);
        }else if (res.statusCode==301 || res.statusCode==302){
            var nextTarget = res.headers.location;
            process.nextTick(function(){
                httpRequest(nextTarget, callback, data, headers);
            });
        }else{
            callback && callback(null, 
                "Server return " + res.statusCode + ": " + buffer);
        }
    });
}

function cookieToStr(cookie){
    if (cookie){
        var result = "";
        for (var key in cookie){
            result += key + (cookie[key] ? "=" + cookie[key] : "") + "; ";
        }
        return result;
    }
    return "";
}

function setCookieByHttpRes(res){
    var rawCookie = res.headers["set-cookie"];
    rawCookie && rawCookie.forEach(function(line){
        line && line.split(';').forEach(function(c){
            var parts = c.split('=');
            _cookie[parts[0].trim()] = (parts[1] || '').trim();
        });
    });
}

function genRequestOptions(rawUrl, errCallback, data, headers){
    var target;
    try{
        target = url.parse(rawUrl);
    }catch(err){
        errCallback("URL parse error, please input validate URL");
        return;
    }
    if (target && target.host){
        target.host = target.port ? 
            target.host.replace(":" + target.port, "") : target.host;
        var result = {
            path: rawUrl,
            method: data ? "POST" : "GET",
            host: target.host, // will be replaced later
            headers: {
                Host: target.host,
                "User-Agent": "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:13.0) Gecko/20100101 Firefox/13.0",
                Connection: "keep-alive",
                Accept: "text/html, text/javascript, application/json, application/xhtml+xml, application/xml;q=0.9, */*;q=0.8",
                Cookie: cookieToStr(_cookie)
            }
        };
        // apply content length
        if (data){
            result.headers["Content-Length"] = Buffer.byteLength(data);
        }
        // apply custom header
        if (headers){
            for (var key in headers){
                result.headers[key] = headers[key];
            }
        }
        var proxy = getProxy();
        if (target.protocol==="http:"){
            result.port = target.port ? target.port : 80;
            result._protocol = "http";
            if (proxy){
                result.host = proxy.host;
                result.port = proxy.port;
            }
        }else if (target.protocol==="https:"){
            result.port = target.port ? target.port : 443;
            result._protocol = "https";
            if (proxy){
                result.agent = tunnel.httpsOverHttp({
                    proxy:{
                        host: proxy.host,
                        port: proxy.port
                    }
                });
            }
        }else{
            errCallback("Not support URL scheme, only support http and https");
            return;
        }
        return result;
    }
}

function getProxy(){
    var proxy;
    if (_proxy){
        proxy = _proxy; 
    }else if (process.env.http_proxy){
        proxy = process.env.http_proxy;
    }
    proxy = proxy ? url.parse(proxy) : null;
    if (proxy && proxy.host){
        return {
            host: proxy.port ? proxy.host.replace(":"+proxy.port, "") : proxy.host,
            port: proxy.port ? proxy.port : "80"
        };
    }
    return;
}
